data_warehouse/services/coverage_statistics_service.py

```diff
@@ -0,0 +1,208 @@
+
+    @staticmethod
+    def groupby_departments(users_qs):
+        '''按照部门分组
+
+        Parameters
+        -------
+        users_qs: QuerySet
+            待分组的User查询集
+
+        Returns
+        ------
+        list of dict
+            按部门后分组后的结果，list包含的元素dict定义为： {
+                department: string,
+                coverage_count: int,
+                    参数users_qs包含的用户中，隶属该部门的用户人数。
+                total_count: int
+                    系统里隶属该部门的用户总人数。
+            }
+        '''
+        data = []
+        coverage_departments = users_qs.values('department__name').annotate(
+            coverage_count=Count('department__name'))
+        for item in coverage_departments:
+            data.append(
+                {
+                    'department': item['department__name'],
+                    'coverage_count': item['coverage_count'],
+                    'total_count': len(User.objects.filter(
+                        department__name=item['department__name']))
+                }
+            )
+        return data
+
+    @staticmethod
+    def get_traning_records(user, program_id=None, department_id=None,
+                            start_time=None, end_time=None):
+        '''根据项目ID和起始时间查询所有的培训活动的所有的培训记录（包含校内和校外活动)
+
+        Parameters
+        ----
+        user: User
+            执行此操作的用户
+
+        program_id: int
+            培训项目ID，为None表示查询所有的培训项目。
+
+        department_id: int
+            需要导出培训项目所属的部门，非校级管理员必须给此参数，校级管理员为此参数None则导出所有院系的所有培训项目。
+
+        start_time: datetime
+            培训活动开始时间
+
+        end_time: datetime
+            培训活动结束时间
+
+        Return
+        ------
+
+        QuerySet
+            包含Record的查询集
+        '''
+        department_ids = []
+        if user is None or not (user.is_department_admin or user
+                                .is_school_admin):
+            raise BadRequest('你不是管理员，无权操作。')
+
+        if department_id is not None:
+            department = Department.objects.filter(id=department_id)
+            if not department:
+                raise BadRequest('部门不存在。')
+            if user.check_department_admin(department[0]) is False:
+                raise BadRequest('你不是该院系的管理员，无权操作。')
+            department_ids.append(department_id)
+        else:
+            if user.is_school_admin:
+                department_ids.extend([
+                    dept.id for dept in DepartmentService
+                    .get_top_level_departments()])
+            else:
+                raise BadRequest('你不是校级管理员，必须指定部门ID。')
+
+        if end_time is None:
+            end_time = now()
+        if start_time is None:
+            start_time = end_time.replace(month=1, day=1, hour=0, minute=0,
+                                          second=0)
+
+        if program_id is not None:
+            program = Program.objects.filter(id=program_id)
+            if not program:
+                raise BadRequest('培训项目不存在。')
+            if program[0].department.id not in department_ids:
```

Comment:
Use `program[0].department_id` instead, it's the real data stored in `Program` table, use `program[0].department.id` would cause another sql lookup.

Reply:
```python
if program_id is not None:
    program = Program.objects.filter(id=program_id)
    if not program:
        raise BadRequest('培训项目不存在。')
    if program[0].department_id not in department_ids:
        raise BadRequest
    program_ids = [program.id]
else:
    program_ids = list(Program.objects.filter(department_id__in=department_ids).values_list('id', flat=True))
# 1. Query for campus event records with program_ids and other conditions.
# 2. Query for off campus event records if program_id is not None, otherwise off_campus_records = Record.objects.none()
# 3. Union querysets
```
```